# Collaborative Filtering Recommendation system

### Imports

In [2]:
import numpy as np
import pandas as pd
from sklearn.decomposition import TruncatedSVD

The data for this project is currently stored as two seperate csv files. I will use pandas to load them as two seperate dataframes and combine them by the "book id" column

In [6]:
ratings_file = "ratings.csv"
ratings_df = pd.read_csv(ratings_file)

books_file = "books.csv"
books_df = pd.read_csv(books_file)

combined_df = pd.merge(ratings_df, books_df, on="book_id")
combined_df.head()

book_id  user_id  rating  id  best_book_id   work_id  books_count  \
0        1      314       5  27             1  41335427          275   
1        1      439       3  27             1  41335427          275   
2        1      588       5  27             1  41335427          275   
3        1     1169       4  27             1  41335427          275   
4        1     1185       4  27             1  41335427          275   

        isbn        isbn13                      authors  ...  ratings_count  \
0  439785960  9.780440e+12  J.K. Rowling, Mary GrandPré  ...        1678823   
1  439785960  9.780440e+12  J.K. Rowling, Mary GrandPré  ...        1678823   
2  439785960  9.780440e+12  J.K. Rowling, Mary GrandPré  ...        1678823   
3  439785960  9.780440e+12  J.K. Rowling, Mary GrandPré  ...        1678823   
4  439785960  9.780440e+12  J.K. Rowling, Mary GrandPré  ...        1678823   

  work_ratings_count work_text_reviews_count ratings_1  ratings_2  ratings_3  \
0            1785676                   27520      7308      21516     136333   
1            1785676                   27520      7308      21516     136333   
2            1785676                   27520      7308      21516     136333   
3            1785676                   27520      7308      21516     136333   
4            1785676                   27520      7308      21516     136333   

   ratings_4  ratings_5                                          image_url  \
0     459028    1161491  https://images.gr-assets.com/books/1361039191m...   
1     459028    1161491  https://images.gr-assets.com/books/1361039191m...   
2     459028    1161491  https://images.gr-assets.com/books/1361039191m...   
3     459028    1161491  https://images.gr-assets.com/books/1361039191m...   
4     459028    1161491  https://images.gr-assets.com/books/1361039191m...   

                                     small_image_url  
0  https://images.gr-assets.com/books/1361039191s...  
1  https://images.gr-assets.com/books/1361039191s...  
2  https://images.gr-assets.com/books/1361039191s...  
3  https://images.gr-assets.com/books/1361039191s...  
4  https://images.gr-assets.com/books/1361039191s...  

[5 rows x 25 columns]

The next step is to pivot the table so that the book names are the columns and the rows are the users indivdual book ratings. In this step I also give any books that a user has not reviewed a socre of 0.

In [10]:
rating_utility_matrix = combined_df.pivot_table(values="rating", index="user_id", columns="title", fill_value=0)
rating_utility_matrix.head()

title    'Salem's Lot  'Tis (Frank McCourt, #2)  \
user_id                                           
2                   0                         0   
3                   0                         0   
4                   0                         0   
7                   0                         0   
9                   0                         0   

title    1421: The Year China Discovered America  1776  1984  \
user_id                                                        
2                                              0     0     0   
3                                              0     0     0   
4                                              0     0     0   
7                                              0     0     0   
9                                              0     0     0   

title    A Bend in the River  A Bend in the Road  A Brief History of Time  \
user_id                                                                     
2                          0                   0                        0   
3                          0                   0                        0   
4                          0                   0                        0   
7                          0                   0                        0   
9                          0                   0                        0   

title    A Briefer History of Time  A Case of Need  ...  \
user_id                                             ...   
2                                0               0  ...   
3                                0               0  ...   
4                                0               0  ...   
7                                0               0  ...   
9                                0               0  ...   

title    Women in Love (Brangwen Family, #2)  \
user_id                                        
2                                          0   
3                                          0   
4                                          0   
7                                          0   
9                                          0   

title    World War Z: An Oral History of the Zombie War  \
user_id                                                   
2                                                     0   
3                                                     0   
4                                                     0   
7                                                     0   
9                                                     0   

title    World Without End (The Kingsbridge Series, #2)  Wuthering Heights  \
user_id                                                                      
2                                                     0                  0   
3                                                     0                  0   
4                                                     0                  0   
7                                                     0                  0   
9                                                     0                  0   

title    Xenocide (Ender's Saga, #3)  Year of Wonders  \
user_id                                                 
2                                  0                0   
3                                  0                0   
4                                  0                0   
7                                  0                0   
9                                  0                0   

title    You Shall Know Our Velocity!  \
user_id                                 
2                                   0   
3                                   0   
4                                   0   
7                                   0   
9                                   0   

title    Zen and the Art of Motorcycle Maintenance: An Inquiry Into Values  \
user_id                                                                      
2                                                        0                   
3                  

In [11]:
rating_utility_matrix.shape

(28906, 812)

Here we can see there are 28,906 different customers. To find a more generalized look at customers reviews I'm going to use the TruncatedSVD iimport from sklearn. 

In [12]:
X = rating_utility_matrix.T
SVD = TruncatedSVD(n_components=30)
transposed_matrix = SVD.fit_transform(X)
transposed_matrix.shape

(812, 30)

Printing the shape of the transposed matrix shows that the original 28,906 rows fort the cusomers have now been compressed in to just 30.

The last step of this project is to make a function to recommend books based upon the input of a book title

In [13]:
corr_matrix = np.corrcoef(transposed_matrix)

In [42]:
def recommend(title,similarity):
    books = rating_utility_matrix.columns
    books_list = list(books)
    book_index = books_list.index(title)
    corr_title = corr_matrix[book_index]
    recommendations = list(books[(corr_title < 1.0) & (corr_title > similarity)])
    return recommendations

In [46]:
title = "Wuthering Heights"  #Change this to alter the starting book 
similarity = 0.8             #Change this to alter the amount of recommedations

print(recommend(title, similarity))

['Cane River', 'Night (The Night Trilogy #1)', 'The Power and the Glory', 'Wuthering Heights']


# Conclusion

In this note book I took the book and customer data from goodread and combined them to create on larger datafram that I them used sklearns SVD to reduce the di